STM8 registers

tyalie opened the pull request:
> [!NOTE]
> I know that there's a branch called stm8-register and this one is called stm8-registers. When we created the stm8-register branch, I don't think we very ready to actually separate every feature in their own branch as the fundamental structure was missing and we did not knew enough. But that has changed, so I'll try again

 This PR tracks what the title says. This is for implementing / feature complete the `STM8RegisterInfo` class and the accompanying `STM8RegisterInfo.td` file.

The current status is that there are a few TODOs in these two files that still need to be looked at / two functions that need to be fully implemented. I'll track the TODOs here too.

## TODOs
- [ ] are the chosen register classes sensible?
- [ ] is `R_PCHL` necessary?
- [ ] `STM8RegisterInfo::getPointerRegClass`
- [ ] `STM8RegisterInfo::eliminateFrameIndex`

### are the chosen reg classes sensible?
This is a TODO mostly to remind us that this can be easily modified if necessary. The current register classes chosen mostly orient themselves according how the AVR backend does it, but it could also mean that it doesn't fit for STM8. 

We will probably discover problems here only over time as we implement other aspects of the backend - so I expect this to be open for a long time.


### is `R_PCHL` necessary?
This is a register that joins the PC High and Low register into a single 16bit reg. The description from the previously TODO also applies here. We need to figure this out over a longer time period.

### `STM8RegisterInfo::getPointerRegClass`
Every backend must override this method. For other targets this might be quite easy, but several comments in the LLVM codebase suggests that this can be instruction depended. In some cases that could mean that this depends on the addressing mode we're operating in. Currently just the PC reg class was chosen.

### `STM8RegisterInfo::eliminateFrameIndex`
Yeah no clue for now. I haven't rly took the time to look into it yet

tyalie committed:
[wip] stm8: Improve STM8RegistersInfo class

This class was previously only a stub and as such did not implement any
useful changes nor did it implement everything that was required for it
not to be an abstract class.

With this all at least required methods are implement as stubs, and the
trivial ones also have a valid implementation now.

What is notably missing here is:
- elimateFrameIndex: this is something I haven't yet bothered to
  understand. But according to AVR it does something with a stack I
  guess
- getPointerRegClass: we have different pointer sizes regarding the
  addressing mode. This might have an influence here

Signed-off-by: Sophie 'Tyalie' Friedrich <dev@flowerpot.me>

nrdmn commented on llvm/lib/Target/STM8/STM8RegisterInfo.cpp:
```diff
@@ -27,4 +27,49 @@
+BitVector STM8RegisterInfo::getReservedRegs(const MachineFunction &MF) const {
+  BitVector Reserved(getNumRegs());
+
+  Reserved.set(STM8::R_PC);
+  Reserved.set(STM8::R_PCE);
+  Reserved.set(STM8::R_PCH);
+  Reserved.set(STM8::R_PCL);
+  Reserved.set(STM8::R_PCHL);
```
I just noticed that AVR doesn't mark PC as reserved (and in fact, it doesn't even define it at all in AVRRegisterInfo.td!). This is probably because on AVR there's no way to directly interact with PC except through branch instructions. You'll never see it as an instruction operand. This is also true for STM8 so we might be able to get away with not defining PC at all.

tyalie replied:
Can you give me an example for such an instruction in e.g. ARM or X86?

Edit: Okay I see what you mean. Take for example the `ADD` instruction in AARCH32 which is defined as `ADD{S}{cond} {Rd}, Rn, Operand2`. And I guess what you meant here is that `Rd` can be the PC register (with a few exceptions).

tyalie committed:
STM8: Remove PC register definition

It was noticed, that AVR does not reserve nor define the PC register in
its backend (see [the discussion][dis]). The current theory is, that due
to the fact that one cannot directly interact with the PC register -
except through specialised instructions (jump, …) - it is not necessary
to define it as an register, which would be used internally by LLVM to
describe mnemonic operands.

We also noticed adapting the definitions, that `getPointerRegClass` is
used to describe registers that can hold pointer values, instead of it
describing the PC register, as assumed before. As our ISA contains the
instructions JP (X) or JP (Y), whereby X and Y hold the pointer, the
previous PC reg class was replaced by a class containing these two.

[dis]: https://github.com/nrdmn/llvm-project/pull/5#discussion_r1538079279

Signed-off-by: Sophie 'Tyalie' Friedrich <dev@flowerpot.me>

nrdmn commented on llvm/lib/Target/STM8/STM8RegisterInfo.cpp:
```diff
@@ -27,4 +27,49 @@
 
 namespace llvm {
 
+/// Initialize register info
+/// as there's no return address register, we Initialize
+/// that with 0
+STM8RegisterInfo::STM8RegisterInfo() : STM8GenRegisterInfo(0) {}
```
Some architectures, e.g. x86 and MSP430, set the program counter as their return address register. What's up with that?

nrdmn commented on llvm/lib/Target/STM8/STM8RegisterInfo.cpp:
```diff
@@ -27,4 +27,49 @@
+                                     unsigned Kind) const {
+  // TODO: extend to actually support different memory models, so we should look
+  // at machine function
+  return &STM8::GPRPCERegClass;
```
Where is `GPRPCERegClass` defined?

nrdmn commented on llvm/lib/Target/STM8/STM8RegisterInfo.td:
```diff
@@ -72,16 +60,12 @@ def GPR8 : RegisterClass<"STM8", [i8], 8, (
   add R_A,
   // index registers
   R_XL, R_XH, R_YL, R_YH,
-  // program counter
-  R_PCL, R_PCH, R_PCE
```
This won't compile anymore. You'll have to remove the trailing commas, i.e.
```diff
-R_XL, R_XH, R_YL, R_YH,
+R_XL, R_XH, R_YL, R_YH
```
